Foundation/JSONSerialization.swift

```diff
@@ -28,6 +28,253 @@ extension JSONSerialization {
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        try visitor.visit(self)
+    }
+}
+
+//@_specialize(where T == Double)
+//@_specialize(where T == Float)
+//extension<T: FloatingPoint> T : _JSONValue {
+extension Float : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        guard isFinite && !isNaN else {
+            throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListReadCorrupt.rawValue, userInfo: ["NSDebugDescription" : "Invalid number value (\(self)) in JSON write"])
+        }
+        try visitor.visit(self)
+    }
+}
+
+//TODO: delete this, use parameterized extensions
+extension Double : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        guard isFinite && !isNaN else {
+            throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListReadCorrupt.rawValue, userInfo: ["NSDebugDescription" : "Invalid number value (\(self)) in JSON write"])
+        }
+        try visitor.visit(self)
+    }
+}
+
+extension Bool : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        try visitor.visit(self)
+    }
+}
+
+extension Decimal : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        guard isFinite else {
+            throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListReadCorrupt.rawValue, userInfo: ["NSDebugDescription" : "Invalid number value (\(self)) in JSON write"])
+        }
+        try visitor.visit(self)
+    }
+}
+
+extension NSNull : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        try visitor.visit(self)
+    }
+}
+
+extension NSNumber : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        if CFNumberIsFloatType(_cfObject) {
+            try doubleValue.visit(with: &visitor)
+        } else {
+            switch _cfTypeID {
+            case CFBooleanGetTypeID():
+                try boolValue.visit(with: &visitor)
+            default:
+                try intValue.visit(with: &visitor)
+            }
+        }
+    }
+}
+
+extension NSArray : _JSONValue {
```

Comment:
This should actually be a big win. No need to bridge NSArrays and NSDictionaries at all when you can just iterate through them directly